// @amk-7/fetch/dist/services/ai/gemini.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_fs_1 = __importDefault(require("node:fs"));
const string_utils_1 = require("../../utils/string.utils");
const genai_1 = require("@google/genai");
const mime_1 = __importDefault(require("mime"));
class GeminiProvider {
    constructor(geminiAPIKey) {
        this.gemini = null;
        this.gemini = new genai_1.GoogleGenAI({
            apiKey: geminiAPIKey,
        });
    }
    static getProvider(geminiAPIKey) {
        if (GeminiProvider.instance == null) {
            GeminiProvider.instance = new GeminiProvider(geminiAPIKey);
        }
        return GeminiProvider.instance;
    }
    async textToText(prompt) {
        if (!this.gemini) {
            throw new Error("Gemini client is not initialized.");
        }
        const response = await this.gemini.models.generateContent({
            model: "gemini-2.5-flash",
            contents: prompt,
        });
        return (0, string_utils_1.removeJsonAnotaionOnString)(response.text);
    }
    saveBinaryFile(fileName, content) {
        node_fs_1.default.writeFile(fileName, content, (err) => {
            if (err) {
                console.error(`❌ Error writing file ${fileName}:`, err);
                return;
            }
            console.log(`✅ File ${fileName} saved to file system.`);
        });
    }
    async textToTextAndImage(prompt) {
        if (!this.gemini) {
            throw new Error("Gemini client is not initialized.");
        }
        const response = await this.gemini.models.generateContentStream({
            model: "gemini-2.5-flash",
            config: {
                responseModalities: ["TEXT"],
            },
            contents: [
                {
                    role: "user",
                    parts: [{ text: prompt }],
                },
            ],
        });
        let fileIndex = 0;
        for await (const chunk of response) {
            if (!chunk.candidates?.[0]?.content?.parts) {
                continue;
            }
            const parts = chunk.candidates[0].content.parts;
            for (const part of parts) {
                if (part.text) {
                    console.log("📝 Texte généré:", part.text);
                }
                else if (part.inlineData) {
                    const fileName = `gemini-image-${fileIndex++}`;
                    const fileExtension = mime_1.default.getExtension(part.inlineData.mimeType || "") || "png";
                    const buffer = Buffer.from(part.inlineData.data || "", "base64");
                    this.saveBinaryFile(`${fileName}.${fileExtension}`, buffer);
                }
            }
        }
        return;
    }
}
exports.default = GeminiProvider;
GeminiProvider.instance = null;
