// @amk-7/fetch/dist/core/jokes/jokeAction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const joke_1 = __importDefault(require("./joke"));
const utils_1 = require("./utils");
const internet_utils_1 = require("../../utils/internet.utils");
const gemini_1 = __importDefault(require("../../services/ai/gemini"));
class JokeAction {
    constructor(geminiAPIKey) {
        this.ai = null;
        if (geminiAPIKey) {
            this.ai = gemini_1.default.getProvider(geminiAPIKey);
        }
    }
    static getAction(geminiAPIKey) {
        if (JokeAction.instance == null) {
            JokeAction.instance = new JokeAction(geminiAPIKey);
        }
        return JokeAction.instance;
    }
    static async getLocalJoke() {
        let yield_content = (await this.jokesGen.next());
        if (yield_content.done) {
            this.jokesGen = (0, utils_1.jokesGenerator)(this.localJokesFilePath);
            yield_content = (await this.jokesGen.next());
        }
        return yield_content.value;
    }
    async generate(options) {
        try {
            const online = await (0, internet_utils_1.checkConnection)();
            if (!online || !this.ai) {
                return (0, utils_1.getRandomJoke)(JokeAction.localJokesFilePath);
            }
            const prompt = joke_1.default.getPrompt(options);
            const jokeData = await this.ai.textToText(prompt);
            const joke = joke_1.default.fromJSON(JSON.parse(jokeData));
            (0, utils_1.saveJoke)(JokeAction.localJokesFilePath, joke);
            return joke;
        }
        catch (error) {
            throw error;
        }
    }
}
exports.default = JokeAction;
JokeAction.localJokesFilePath = node_path_1.default.join(__dirname, "data.json");
JokeAction.jokesGen = (0, utils_1.jokesGenerator)(JokeAction.localJokesFilePath);
JokeAction.instance = null;
