// @amk-7/fetch/dist/core/jokes/joke.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Joke {
    constructor(setup, punchline, options) {
        this.setup = setup;
        this.punchline = punchline;
        if (options) {
            this.options = options;
        }
    }
    getSetup() {
        return this.setup;
    }
    getPunchline() {
        return this.punchline;
    }
    getOptions() {
        return this.options;
    }
    toString() {
        return `${this.setup} - ${this.punchline}`;
    }
    static fromJSON(json) {
        return new Joke(json.setup, json.punchline, json.options);
    }
    toJSON() {
        return {
            setup: this.setup,
            punchline: this.punchline,
            options: this.options,
        };
    }
    static getPrompt(options) {
        let prompt = "Génère une blague courte et drôle";
        if (options?.theme) {
            prompt += ` sur le thème "${options.theme}"`;
        }
        prompt += `. Réponds uniquement en JSON valide avec la structure suivante : 
        {
            "setup": "<texte de la mise en place>",
            "punchline": "<texte de la chute>",
            "options": {
                "theme": "<thème si défini>"
            }
        }`;
        return prompt;
    }
}
exports.default = Joke;
// console.log((Joke.prototype as any).__props);
