// @amk-7/fetch/dist/core/memes/memeAction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_path_1 = __importDefault(require("node:path"));
const meme_1 = __importDefault(require("./meme"));
const internet_utils_1 = require("../../utils/internet.utils");
const gemini_1 = __importDefault(require("../../services/ai/gemini"));
const logger_1 = __importDefault(require("../../utils/logger"));
class MemeAction {
    constructor() {
        this.ai = null;
    }
    static getAction() {
        if (MemeAction.instance == null) {
            MemeAction.instance = new MemeAction();
        }
        return MemeAction.instance;
    }
    static async getLocalMeme() {
        return new meme_1.default(1, "Titre par défaut", "http://example.com/meme.jpg", {});
        // let yield_content = (await this.MemesGen.next());
        // if (yield_content.done){
        //     this.MemesGen = MemesGenerator(this.localMemesFilePath);
        //     yield_content = (await this.MemesGen.next());
        // }
        // return yield_content.value as Meme;
    }
    getAI() {
        if (!this.ai) {
            this.ai = gemini_1.default.getProvider();
        }
        return this.ai;
    }
    async generate(options) {
        try {
            logger_1.default.debug("Vérification connexion...");
            const online = await (0, internet_utils_1.checkConnection)();
            if (!online) {
                logger_1.default.warn("Pas de connexion Internet. Utilisation du fallback local.");
                throw new Error("No internet connection.");
            }
            logger_1.default.info("Connexion OK, génération via IA");
            const prompt = meme_1.default.getJsonPrompt(options);
            const MemeData = await this.getAI().textToText(prompt);
            return meme_1.default.fromJSON(JSON.parse(MemeData));
        }
        catch (error) {
            logger_1.default.error(error);
            throw new Error("Meme generation failed.");
            //return getRandomMeme(MemeAction.localMemesFilePath);
        }
    }
    async generateMediaAndJson(options) {
        try {
            logger_1.default.debug("Vérification connexion...");
            const online = await (0, internet_utils_1.checkConnection)();
            if (!online) {
                logger_1.default.warn("Pas de connexion Internet. Utilisation du fallback local.");
                throw new Error("No internet connection.");
            }
            logger_1.default.info("Connexion OK, génération media via IA");
            const prompt = meme_1.default.getMixedPrompt(options);
            const MemeData = await this.getAI().textToTextAndImage(prompt);
            return meme_1.default.fromJSON({ id: 1, title: "Titre par défaut", imageUrl: "", options: options });
        }
        catch (error) {
            logger_1.default.error(error);
            throw new Error("Meme media generation failed.");
            //return getRandomMeme(MemeAction.localMemesFilePath);
        }
    }
}
exports.default = MemeAction;
MemeAction.localMemesFilePath = node_path_1.default.join(__dirname, "default.json");
MemeAction.instance = null;
