// @amk-7/fetch/dist/core/memes/meme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Meme {
    constructor(id, title, imageUrl, options) {
        this.id = id;
        this.title = title;
        this.imageUrl = imageUrl;
        this.options = options;
    }
    getOptions() {
        return this.options;
    }
    toString() {
        return `${this.title} -> ${this.imageUrl}`;
    }
    static fromJSON(json) {
        return new Meme(json.id, json.title, json.imageUrl, json.options);
    }
    static toJSON(meme) {
        return {
            id: meme.id,
            title: meme.title,
            imageUrl: meme.imageUrl,
            options: meme.options,
        };
    }
    static getPrompt(options) {
        let prompt = "Génère un meme drôle";
        if (options.mood)
            prompt += ` sur le mood (mood: string) "${options.mood}"`;
        if (options.characters?.length)
            prompt += ` avec  ${options.characters.join(", ")} (characters: string[])`;
        if (options.topText)
            prompt += ` avec le texte en haut (topText: string) "${options.topText}"`;
        if (options.bottomText)
            prompt += ` avec le texte en bas (bottomText: string) "${options.bottomText}"`;
        if (options.style)
            prompt += ` dans le style (style: string) "${options.style}"`;
        if (options.format)
            prompt += ` au format (format: string) "${options.format}"`;
        if (options.language)
            prompt += ` en langue (language: string) "${options.language}"`;
        if (options.nsfw !== undefined)
            prompt += ` et que ce soit ${options.nsfw ? "" : "pas "} NSFW (nsfw: boolean)`;
        if (options.customImageUrl)
            prompt += ` en utilisant cette image de base (customImageUrl: string) "${options.customImageUrl}"`;
        return prompt;
    }
    static getMediaPrompt(options) {
        let prompt = this.getPrompt(options);
        prompt += `. Réponds uniquement par l'URL de l'image générée.`;
        return prompt;
    }
    static getJsonPrompt(options) {
        let prompt = this.getPrompt(options);
        prompt += `. Réponds uniquement en JSON valide avec la structure suivante : 
        {
            "id": <number>,
            "title": "<titre du meme>",
            "options": {
                "mood": "<mood si défini>",
                "format": "<format si défini>",
                "style": "<style si défini>",
                "topText": "<topText si défini>",
                "bottomText": "<bottomText si défini>",
                "language": "<language si défini>",
                "nsfw": <nsfw si défini>,
                "characters": <characters si défini>
            }
        }`;
        return prompt;
    }
    static getMixedPrompt(options) {
        let prompt = this.getPrompt(options);
        prompt += `. Génére moi l'image et donne moi le JSON valide avec la structure suivante correspondant au meme généré : 
        {
            "id": <number>,
            "title": "<titre du meme>",
            "options": {
                "mood": "<mood si défini>",
                "format": "<format si défini>",
                "style": "<style si défini>",
                "topText": "<topText si défini>",
                "bottomText": "<bottomText si défini>",
                "language": "<language si défini>",
                "nsfw": <nsfw si défini>,
                "characters": <characters si défini>
            }
        }`;
        return prompt;
    }
}
exports.default = Meme;
